internal/core/src/segcore/SegmentSealed.h

```diff
@@ -73,63 +83,69 @@ class SegmentSealed : public SegmentInternalInterface {
+                if (index.field_id != field_id) {
+                    return false;
+                }
+                if (index.cast_type.data_type() ==
+                    JsonCastType::DataType::JSON) {
+                    // for json flat index, path should be a subpath of nested_path
+                    return path.substr(0, index.nested_path.length()) ==
+                           index.nested_path;
+                }
+                if (any_type) {
+                    return true;
+                }
+                return index.nested_path == path &&
+                       index.index->IsDataTypeSupported(data_type);
+            });
+        return it != json_indices.end();
     }
 
  protected:
-    struct JSONIndexKey {
+    virtual PinWrapper<const index::IndexBase*>
+    chunk_index_impl(FieldId field_id, int64_t chunk_id) const override = 0;
+
+    PinWrapper<const index::IndexBase*>
+    chunk_index_impl(FieldId field_id,
+                     std::string path,
+                     int64_t chunk_id) const override {
+        auto it = std::find_if(json_indices.begin(),
+                               json_indices.end(),
+                               [field_id, path](const JsonIndex& index) {
+                                   if (index.field_id != field_id) {
+                                       return false;
+                                   }
+                                   if (index.cast_type.data_type() !=
+                                       JsonCastType::DataType::JSON) {
+                                       // json path index
+                                       return index.nested_path == path;
+                                   }
+                                   // json flat index
+                                   return path.substr(
+                                              0, index.nested_path.length()) ==
+                                          index.nested_path;
+                               });
+        return it != json_indices.end() ? it->index.get() : nullptr;
+    }
+    struct JsonIndex {
         FieldId field_id;
         std::string nested_path;
-        bool
-        operator==(const JSONIndexKey& other) const {
-            return field_id == other.field_id &&
-                   nested_path == other.nested_path;
-        }
+        JsonCastType cast_type{JsonCastType::UNKNOWN};
+        index::IndexBasePtr index;
     };
 
-    struct hash_helper {
-        size_t
-        operator()(const JSONIndexKey& k) const {
-            std::hash<int64_t> h1;
-            std::hash<std::string> h2;
-            size_t hash_result = 0;
-            boost::hash_combine(hash_result, h1(k.field_id.get()));
-            boost::hash_combine(hash_result, h2(k.nested_path));
-            return hash_result;
-        }
-    };
-    std::unordered_map<JSONIndexKey, index::IndexBasePtr, hash_helper>
-        json_indexings_;
+    std::vector<JsonIndex> json_indices;
```

Comment:
seems we changed to use vector instead of unordered_map because for flat index the path is matched as substr. but could this lead to poor performance if we have lots of path indexes?

would it make sense to use a vector for flat index and a unordered_map for path index? will this add too much complexity to impl?

Reply:
as discussed offline: this may cause performance issue if hundreds of thousands path/flat indexes are created on a single JSON field, but we can delay the optimization till we see it.

Reply:
> as discussed offline: this may cause performance issue if hundreds of thousands path/flat indexes are created on a single JSON field, but we can delay the optimization till we see it.

agree